### PR TITLE
JBPM-5813: Stunner - Open XML editor instead of sorry screen when process can't be opened

### PR DESCRIPTION
**Pull request by manstis**
See https://issues.jboss.org/browse/JBPM-5813

BPMN and CM now support opening the XML in a XML text editor if it failed parsing on the server. You can test by creating (for example) a new process file, cloning the workbench repository, adding a broken file, commit and push. When you open the broken file the editor will be a XML text editor. Save, versioning etc works as normal.

This depends on https://issues.jboss.org/browse/JBPM-5813 being merged (then I'll rebase to 1 commit).

Only the third commit needs reviewing for this PR (or, leave until the above is merged).

### CODE REVIEW

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java [hunk 1]
```diff
@@ -256,6 +269,106 @@ protected void open(final ProjectDiagram diagram) {
...
+            }
+            return hash;
+        });
+
+        return proxy;
+    }
+
+    protected ProjectDiagramEditorProxy makeXmlEditorProxy() {
+        final ProjectDiagramEditorProxy proxy = new ProjectDiagramEditorProxy();
+        proxy.setSaveAfterValidationConsumer(Command::execute);
```
Comment (jomarko) [inline kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java]: I am totally not sure what does this line do? `Command::execute` seems to be as reference to interface, isn't reference to some implementation needed?

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java [hunk 2]
```diff
@@ -256,6 +269,106 @@ protected void open(final ProjectDiagram diagram) {
         }
     }
 
+    @SuppressWarnings("unchecked")
+    protected ProjectDiagramEditorProxy makeStunnerEditorProxy() {
```
Comment (jomarko) [inline kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java]: Here I have concern, the method content is quite big, however I found just two tests that verifies `makeStunnerEditorProxy` is called but no test that would verify at least that some handlers are registered on `ProjectDiagramEditorProxy`. If it is not testable I would say than we should choose different and testable approach.

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java (hunk 1 above)
Comment (jomarko) [reply kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java]: And similar concern as for  `makeStunnerEditorProxy`

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java [hunk 3]
```diff
@@ -800,12 +868,33 @@ private void onValidationFailed(final Collection<DiagramElementViolation<RuleVio
         hideLoadingViews();
     }
 
-    private void onLoadError(final ClientRuntimeError error) {
-        showError(error);
+    protected void onLoadError(final ClientRuntimeError error) {
```
Comment (jomarko) [inline kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java]: I think worth for unit test.

**Comment by jomarko** [comment]
When the version history is like
- 1 valid bpmn
- 2 corrupted bpmn

Then if you select version 1 and consequently again version 2, the xml editor is not marked as dirty more.

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-backend/src/test/java/org/kie/workbench/common/stunner/project/backend/service/ProjectDiagramServiceImplTest.java [hunk 4]
```diff
@@ -240,6 +241,30 @@ public void testSave() {
                      result);
     }
 
+    @Test
+    public void testSaveAsXml() {
+        final String xml = "xml";
+        final String comment = "comment";
+        final Map<String, Object> attributes = new HashMap<>();
```
[the remark this answers is not in the document]
Comment (manstis) [reply kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-backend/src/test/java/org/kie/workbench/common/stunner/project/backend/service/ProjectDiagramServiceImplTest.java]: No reason; in value of ```attributes``` makes no real difference; we just ```verify(..)``` that the save used the same object later in the test. I have however updated to use ```Collections.singletonMap("key", "value")``` for consistency.

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java (hunk 1 above)
Comment (manstis) [reply kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java]: It is a Java lamda short-hand "method reference" the expand line is ```proxy.setSaveAfterValidationConsumer((command)->command.execute())``` so in essence the line is invoking the command's execute method. The implementation of the command is define [here](https://github.com/manstis/kie-wb-common/blob/5f8702929c8d582afddc0abf86d640116e15ce96/kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java#L504) and invoked [here](https://github.com/manstis/kie-wb-common/blob/5f8702929c8d582afddc0abf86d640116e15ce96/kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java#L509).

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java (hunk 2 above)
Comment (manstis) [reply kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java]: @jomarko The "proxies" are used by ```AbstractProjectDiagramEditor``` to delegate implementation of various operations to the specific editor (Stunner or XML) - ```hashCode()``` for "is dirty" behaviour, "what to do before the file is saved", "how to save the file", "what to do if there were no changes in the file") [here](https://github.com/manstis/kie-wb-common/blob/5f8702929c8d582afddc0abf86d640116e15ce96/kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java). The implementation of the various "proxied" methods are handled by the various tests relating to "save" in ```AbstractProjectDiagramEditorTest```. That said ```onLoadError(..)``` (mentioned below) shows as not being covered when I run tests with coverage in IDEA, so I will add tests for that scenario.

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java (hunk 3 above)
Comment (manstis) [reply kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java]: @jomarko This is covered by this [test](https://github.com/manstis/kie-wb-common/blob/5f8702929c8d582afddc0abf86d640116e15ce96/kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/test/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditorTest.java#L621) however I will add some more tests for three uncovered failure scenarios (when validation of a Stunner diagram fails, when saving a Stunner diagram fails and when saving the XML fails)... new commit to follow.

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java (hunk 2 above)
Comment (jomarko) [reply kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java]: thanks for explanation

File: kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java (hunk 3 above)
Comment (jomarko) [reply kie-wb-common-stunner/kie-wb-common-stunner-extensions/kie-wb-common-stunner-project/kie-wb-common-stunner-project-client/src/main/java/org/kie/workbench/common/stunner/project/client/editor/AbstractProjectDiagramEditor.java]: thanks

**Comment by manstis** [comment]
@jomarko New commit with more Unit Tests (and corrections to ```attributes``` as requested).

I'll look into other items; but if they are complex I might suggest a new PR @romartin has some "monster" PRs pending and they will be affected by this PR; so I'd prefer to "get this in" and then let @romartin adjust his PRs ;-)

**Comment by manstis** [comment]
@jomarko New commit added to ensure the dock status is correctly refreshed when switching versions (between XML and Stunner). i.e. the dock "preview" and "properties" panels are cleared when viewing the XML text version of a process file (I know you've approved, but I thought I'd let you know it's fixed).

Looking at "is dirty" issue now...

**Comment by manstis** [comment]
@jomarko Regarding "When the version history is like..."

When I open the editor the XML editor is shown; making a change marks the editor as dirty and the "lock" icon appears in both the Project Explorer and the editor title. If I select a different version the "lock" icon remains in the Project Explorer but disappears in the editor title. If I again select the latest version the "lock" icon remains in the Project Explorer and remains disappeared in the editor title. If I then edit the file the lock does not reappear in the editor title.

Is this what you meant?

This is an issue with how Stunner is updating the editor title (the "lock" icon should not disappear when selecting different versions - or at least that'd be consistent with other editors - although that in itself seems wrong - if I switch versions IMO the lock should be released until I re-select the "editable" version and attempt to make a change).

**Pull request merged**